resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSetManager.scala

```diff
@@ -52,12 +56,28 @@ private[spark] class KubernetesTaskSetManager(
           if (pendingTasksClusterNodeIP.nonEmpty) {
             logDebug(s"Got preferred task list $pendingTasksClusterNodeIP for executor host " +
               s"$executorIP using cluster node IP $clusterNodeIP")
+            pendingTasksClusterNodeIP
+          } else {
+            val clusterNodeFullName = inetAddressUtil.getFullHostName(clusterNodeIP)
+            val pendingTasksClusterNodeFullName = super.getPendingTasksForHost(clusterNodeFullName)
+            if (pendingTasksClusterNodeFullName.nonEmpty) {
+              logDebug(s"Got preferred task list $pendingTasksClusterNodeFullName " +
+                s"for executor host $executorIP using cluster node full name $clusterNodeFullName")
+            }
+            pendingTasksClusterNodeFullName
           }
-          pendingTasksClusterNodeIP
         }
       } else {
         pendingTasksExecutorIP  // Empty
       }
     }
   }
 }
+
+// To support mocks in unit tests.
+private[kubernetes] class InetAddressUtil {
+
+  def getFullHostName(ipAddress: String): String = {
+    InetAddress.getByName(ipAddress).getCanonicalHostName
```

Comment:
Yes, it does issue a network call to DNS. Caching is done by the `InetAddress` class. From the [java doc](https://docs.oracle.com/javase/7/docs/api/java/net/InetAddress.html):

>InetAddress Caching
>
> The InetAddress class has a cache to store successful as well as unsuccessful host name resolutions.
By default, when a security manager is installed, in order to protect against DNS spoofing attacks, the result of positive host name resolutions are cached forever. When a security manager is not installed, the default behavior is to cache entries for a finite (implementation dependent) period of time. The result of unsuccessful host name resolution is cached for a very short period of time (10 seconds) to improve performance.